### ISSUE TITLE
Implement terminology and Placeholder support for the trainer

### ISSUE THREAD

**Issue by XapaJIaMnu** [hplt-project/OpusTrainer#2]
We need to be able to add terminology support to our trainer. There's multiple ways we can achieve this, but https://aclanthology.org/P19-1294/ is a good starting point.

We already have placeholder generator code, so we should make use of that.

**Comment by ZJaume** [comment]
Bad issue number in the commit?

**Comment by jelmervdl** [comment]
Sorry!

**Comment by XapaJIaMnu** [comment]
I've been investigating this for a couple of days

# Idea 1
The current take:
Take a parallel line and the alignment file and replace random words that have 1-1 alignment with a placeholder token so that the model learns to copy placeholders.

## PROBLEMS:
### Problem 1
When training with guided alignments, we can't replace randomly. We can only replace 1-1 equivalent otherwise training crashes, because the guided alignments wouldn't be accurate.

#### Solution 1: 
Preprocess the corpora with placeholders and then compute alignments.
##### Downsides:
- Always the same position of with the same type of placeholders. I don't like it.

#### Solution 2: 
Check against vocab and only replace tokens which are whole sentencepieces

#### Solution 3: 
Train without guided alignments. What do we use it for? Highlighting. Can we do that without guided alignment?

### Problem 2
Languages that are not tokenized present unique challenges. How do we do placeholding there?
- Chinese/Japanese/Thai/Khmer... -> Run a tokenizer, then fast align, then perform placeholder replacement then detokenize and start training? That's a lot of preprocessing.

### Problem 3
Placeholding morphologically rich languages. We ideally want to have the placeholder only on the uninflected part of the word and have the MT system figure out its inflection after. That would mean that we perform placeholder replacement post SPM, but we normally train on raw text, not sentence split text.

#### Solution
Do APE/Grammatical error correction, that should be trivial enough but it's a 2 pass solution.

I have implemented a proof of concept that can be run by: `./test.py fren.pasted corpus.aln > fren.placeholded.pasted`

I have implemented [proof_of_concept.tar.gz](https://github.com/hplt-project/empty-trainer/files/10503016/proof_of_concept.tar.gz) a solution that takes the fast-align scores of the raw corpora and based on that replaces random words. We should lower the probability of replacement compared to the proof of concept but it seems to work, albeit maybe not the most computationally efficient thing.
```
Page REPLACEMENT4 REPLACEMENT6 1 // 1 Résultats REPLACEMENT2    Page REPLACEMENT4 REPLACEMENT6 1 // 1 Results REPLACEMENT2
La REPLACEMENT3 antimicrobienne est REPLACEMENT2 capacité d'un microorganisme (bactérie, virus ou parasite tel celui de la malaria) à résister à l'action REPLACEMENT6 agent antimicrobien. Antimicrobial REPLACEMENT3 is REPLACEMENT2 ability of a microorganism (e.g., a bacterium, a virus) to resist the action of REPLACEMENT6 antimicrobial agent.
Appliquer       Apply
Remove this filter Type: matériel de REPLACEMENT0 à outils      Remove this filter Type: REPLACEMENT0 material
```
# Idea 2
Add factors for words that we want force translated, ala Dinu et al. That seems easier and doable: https://arxiv.org/abs/1906.01105

@jelmervdl @ZJaume  opinions?

**Comment by ZJaume** [comment]
# Regex without alignment approach
On my side, I've been working with an implementation of terminology support the past two months and we have students already trained for clients and first deployments have been done. The disadvantage of this, is that it was outlined for an scenario where we have a terminology and in-domain data that some of it complies with the terminology. So it might not be usable for empty-trainer. The code is published here: https://github.com/ZJaume/terminology

The approach consists in annotation of the part of in-domain data that both sides of the translation have the correct translation according to the terminology. Terminology is provided by the client in TBX format, we convert it to json dictionary (removing ambiguities because these terminologies are full of them and sometimes quite messy), and use the dictionary to do a regex search and add special token annotation. For each word found in the source (we allow up to 4 extra character suffixes to match inflected words) we check if the target term is found in the target sentence and it's not already annotated. If both sides are found, we annotate them.

The annotation is done by placing special tokens delimiting the term and inserting the target term in the source. It is something like this:
```
Mi pelo es de color <t_start>rosa<t_mid>pink<t_end>.  Mi hair is <t_start>pink<t_end>.
Mis <t_start>apartamento<t_mid>apartment<t_end>s están alquilados.   My <t_start>apartment<t_end>s are rented.
```
This annotation was inspired by https://arxiv.org/pdf/1912.00567.pdf

For inference we generate an [lttoolbox](https://wiki.apertium.org/wiki/Lttoolbox) dictionary and annotate data really fast compared to python regex. Achieving high speed was needed in order to reduce latency of students as much as possible.

## Real-world scenario
We've been using this for two different domains both in German<->English and finetuning the German student of bergamot using the `<misc1>`, `<misc2>` and `<misc3>` special tokens that were included for tagging purposes (see [here](https://www.statmt.org/wmt22/efficiency-task.html)). We still have some human valuation pending and we have to review ourselves carefully the output but, for now, it seems that it works nice. we see terminology compliance of the model increase from 70% to 90% although the general quality degrades. This is due to the fact that the test set is not compliant itself with the terminology, nor data produced by translators. They want the MT engine suggestions to be more meticulous than them and post-edit afterwards if fluency needs to be improved.

This approach has the advantage of not needing alignments and being able to work with multi-words, which, specially in German, are the ones that we are more interested because they are richer and less ambiguous. Furthermore, we've had to ignore less than 4 char words from terminologies because the entries that the clients do, do not have any context information or consider that they could be used in an automated process. For example, we found entries like translating "Zoll" to "inch"  and its abbreviation "in", resulting in terminology injecting "Zoll" in every "in" stop word.

# Opinions on the proposed approaches
Discarding guided alignments in solution3 seems ok to me but raises the doubt about how this transfers to students. Do we also need to drop guided alignments for them? Would this affect their quality? Maybe we don't have to discard guided alignments when distilling a student because synthetic data could be generated using the placeholding/terminolgy script and alignments would be computed on already placeholded data?

Problem 2 agree that it is a lot of processing.

My main concern here is that, will models be able to correctly learn inflections if they don't see the actual words?

If we consider using factors (students don't support it, would be difficult to add support?) and doing the kind of processing that Problem 2 needs, should we consider using [marian example of forced translation](https://github.com/marian-nmt/marian-examples/tree/master/forced-translation) or a modified version?

**Comment by XapaJIaMnu** [comment]
@ZJaume so if i understand correctly, this is a scenario where you have the terminology in advance and it overlaps with some of your training data so you are able to perform this approach via fine tuning.

I agree that potentially getting the inflection right with the blind find/replace would be a problem. Maybe this could be reserved for items that never get inflected (emoji), or languages that don't apply inflection (Chinese/English). 

> My main concern here is that, will models be able to correctly learn inflections if they don't see the actual words? I think that keeping the actual words and inserting target terms like I did in the approach above mentioned, could work better.

This is basically what Dinu et al do, so that should be a default solution.

Re-guided-alignments. According to some industry contacts guided alignments are useless in terms of translation quality, and they have stopped using them in their systems. Looking at Google/M$ Translator, neither is volunteering alignments for their translations so i think they aren't doing them.

Re-Factors: Again according to industry, factors are useless. The same thing can be achieved by appending a word after the word in question `I like <verb> pie.` with `<verb>` being just like that in the sequence achieves the same performance as having a dedicated factor system at a fraction of the computational cost. We can also use them in students. In this manner.
